File: 4_challenge/ghidra_scripts/GhidraDecomp.java
// Ghidra headless script for read-only function listing and decompilation.

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.PrintWriter;

public class GhidraDecomp extends GhidraScript {
    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String queryFile = args.length >= 1 ? args[0] : "/tmp/ghidra_query.txt";
        String outputFile = args.length >= 2 ? args[1] : "/tmp/ghidra_output.txt";

        String query;
        try (BufferedReader reader = new BufferedReader(new FileReader(queryFile))) {
            query = reader.readLine();
        }

        try (PrintWriter writer = new PrintWriter(new FileWriter(outputFile))) {
            if (query == null || query.trim().isEmpty()) {
                writer.println("Empty query file");
                return;
            }

            query = query.trim();
            if ("LIST_FUNCS".equals(query)) {
                FunctionIterator funcs = currentProgram.getFunctionManager().getFunctions(true);
                for (Function func : funcs) {
                    writer.printf("0x%08x  %s%n", func.getEntryPoint().getOffset(), func.getName());
                }
                return;
            }

            Address addr = currentProgram.getAddressFactory().getAddress(query);
            if (addr == null) {
                writer.printf("Invalid address: %s%n", query);
                return;
            }

            Function func = currentProgram.getFunctionManager().getFunctionContaining(addr);
            if (func == null) {
                writer.printf("No function found at address: %s%n", query);
                return;
            }

            DecompInterface decomp = new DecompInterface();
            decomp.openProgram(currentProgram);
            DecompileResults result = decomp.decompileFunction(func, 30, new ConsoleTaskMonitor());

            if (result.decompileCompleted()) {
                writer.print(result.getDecompiledFunction().getC());
            } else {
                writer.printf("Decompilation error: %s%n", result.getErrorMessage());
            }
        }
    }
}

File: 4_challenge/ghidra_scripts/ghidra_decomp.py
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
from java.io import FileReader, FileWriter, BufferedReader, PrintWriter

args = getScriptArgs()
if len(args) >= 2:
    QUERY_FILE = args[0]
    OUTPUT_FILE = args[1]
else:
    QUERY_FILE = "/tmp/ghidra_query.txt"
    OUTPUT_FILE = "/tmp/ghidra_output.txt"

monitor = ConsoleTaskMonitor()

reader = BufferedReader(FileReader(QUERY_FILE))
addr_str = reader.readLine()
reader.close()

if addr_str is None:
    raise RuntimeError("Empty query file")

addr_str = addr_str.strip()

writer = PrintWriter(FileWriter(OUTPUT_FILE))

if addr_str == "LIST_FUNCS":
    fm = currentProgram.getFunctionManager()
    funcs = list(fm.getFunctions(True))
    for func in sorted(funcs, key=lambda f: f.getEntryPoint().getOffset()):
        entry = func.getEntryPoint()
        name = func.getName()
        writer.println("0x%08x  %s" % (entry.getOffset(), name))
else:
    try:
        addr = currentProgram.getAddressFactory().getAddress(addr_str)
    except Exception as e:
        writer.println("Invalid address: %s (%s)" % (addr_str, str(e)))
        writer.close()
        raise RuntimeError("Bad address")

    func = currentProgram.getFunctionManager().getFunctionContaining(addr)

    if func is None:
        writer.println("No function found at address: %s" % addr_str)
    else:
        decomp = DecompInterface()
        decomp.openProgram(currentProgram)
        result = decomp.decompileFunction(func, 30, monitor)

        if result.decompileCompleted():
            c_code = result.getDecompiledFunction().getC()
            writer.print(c_code)
        else:
            writer.println("Decompilation error: %s" % result.getErrorMessage())

writer.close()
